doc/sflow/sflow_hld.md

```diff
@@ -0,0 +1,456 @@
+                                                                                 ; stopped : hsflowd daemon has been stopped
+                                                                                 ; listening : hsflowd is listening for samples
+```
+
+### 6.6 **sflow container**
+
+hsflowd (https://github.com/sflow/host-sflow) is the most popular open source implementation of the sFlow agent and provides support for DNS-SD (http://www.dns-sd.org/) and can be dockerised. hsflowd supports sFlow version 5 (https://sflow.org/sflow_version_5.txt which supersedes RFC 3176). hsflowd will run as a systemd service within the sflow docker.
+
+CLI configurations will be saved to the ConfigDB. Once the genetlink channel has been initialised and the sFlow traps mapped to the genetlink group, the hsflowd service is started. The service startup script will derive the /etc/hsflowd.conf from the ConfigDB. Config changes will necessitate restart of hsflowd. hsflowd provides the necessary statistics for the "show" commands. CLI "config" commands will retrieve the entries in the config DB.
+
+#### sflowmgrd
+
+The sflowmgrd reads the SFLOW_GENETLINK table to create program genetlink groups that will be used by the SAI driver to lift packets to the hsflowd. Once the genetlink group is created, updates the SFLOW_GENETLINK_TABLE table for sFlowOrch to process. A multicast group is typically created which enables multiple applications (e.g. even a local netlink sniffer tools) to additionally inspect the samples that are lifted to hsflowd.
+
+the sflowmgrd daemon listens to SFLOW_COLLECTOR to construct the hsflowd.conf and start the hsflowd service.
+The mapping between the SONiC sflow CLI parameters and the host-sflow is given below:
+
+SONIC CLI parameter| hsflowd.conf equivalent
+-------------------|------------------------
+collector ip-address | collector.ip
+collector port| collector.UDPPort
+agent ip-address | agentIP
+max-datagram-size | datagramBytes
+sample-rate | sampling
+
+The master list of supported host-sflow tokens are found in host-sflow/src/Linux/hsflowtokens.h
+
+Example SONiC CLI configuration:
+
+&#35; sflow collector add sflow-a 10.100.12.13 agent-addr 10.0.0.10 max-datagramsize 1500
+
+&#35; sflow collector add sflow-b 10.144.1.2 pot 6344
+
+&#35; sflow sample-rate 32768
+
+```
+sflow {
+  agentIP = 10.0.0.10
+  sampling = 32748
+  collector { ip = 10.100.12.13 }
+  collector { ip = 10.144.1.2 UDPPort=6344 }
+  datagramBytes = 1500
+}
+```
+
+sflowmgrd also listens to SFLOW to propogate the sampling rate changes to App DB SFLOW_TABLE.
+
+#### hsflowd service
+
+hsflowd provides an module adaptation layer for interfacing with the NOS. In the host-sflow repo, a src/Linux/mod_sonic.c adaption layer will be provided for hsflowd APIs to SONiC that deal with hsflowd initialization, configuration changes, packet sample consumption etc. More specifically, SONiC will register and provide callbacks for the following HSP events:

```

Comment:
What is the configurations need to expose hsflowd/InMon ? If user wanted to expose some custom configuration such as sampling rate what is the strategy, where do maintain the custom configurations? is it in SONiC Repo or hsflowd/InMon Repos?

Reply:
Collector and agent specific configurations are available in the configDB and host-sflow will have a mod_sonic that is customised to interact with SONiC (essentially query the DB). If there are new customizations, we publish the changes in SONiC schema and mod_sonic will have to be updated (in host-sflow repo) to provide the necessary functionality.